### DISCUSSION TITLE
Allow only one response per picture

### DISCUSSION THREAD

**Discussion by avashipman** [jspsych/jsPsych#984] [Q&A] [answered]
I made a plugin that has a visual oddball task with a survey response. The aim of the visual oddball task is for the participant to press the 8 key each time the target image appears (which is often) from an array of other images. I made it so that the data records if the participant responded to the target image by showing null if no response and the trial.choices key (56) if there was a response [(here)](https://github.com/avashipman/Noisy-Channel/blob/f64988fbf29c623e143ca6bff76d10defe6c5aff/jspsych-6.1.0/jspsych-survey-distractor-copy.js#L312-L326). However, if the participant press 8 multiple times then every 8 will be recorded for that target image, which messes up the data. Is there a way to only allow one response for each target image? 

I tried putting the keyboard listener code in the image interval and making [persist false](https://github.com/avashipman/Noisy-Channel/blob/a681534934d33f5913e0d53088c517f510ef7ecb/jspsych-6.1.0/jspsych-survey-distractor-copy.js#L287-L296), thinking that it wouldn't allow additional responses for the image shown in the interval and then it would reset for the next image in the interval. However, this didn't have any effect.

Any ideas/help would be greatly appreciated! Thank you already for all of your help on this code!!

**Comment by vijaymarupudi** [comment, marked as answer]
So what you need to do is add the listener in the changeImage function. And the listener function, once it's called, should remove itself after pushing the appropriate data. This way, only one response per image will be recorded.

Let me know if you have any further questions!

**Reply by avashipman** [reply]
I tried doing that [here](https://github.com/avashipman/Noisy-Channel/blob/a681534934d33f5913e0d53088c517f510ef7ecb/jspsych-6.1.0/jspsych-survey-distractor-copy.js#L270-L298) but it didn't do anything. I can still respond multiple times for each X and each response is recorded. Am I implementing it wrong? Also, the [data](https://github.com/avashipman/Noisy-Channel/blob/a681534934d33f5913e0d53088c517f510ef7ecb/jspsych-6.1.0/jspsych-survey-distractor-copy.js#L322-L336) showing the amount of X's that appeared on the screen is inconsistent. Maybe this is contributing to the listener?

**Reply by vijaymarupudi** [reply]
Just a note, always declare the functions you are using before you use them. In this case, `after_response` should be defined before the `slideShow()`

I'm not exactly sure why that is happening. One thing to check is whether the `after_response` function is running after each keypress by inserting a `console.log` statement there. If it is, it is likely that the listener isn't being removed properly, let's investigate that after you verify this!

**Reply by avashipman** [reply]
It seems like `after_response` is being called every time in the interval, regardless if there is a keypress or not. The log automatically shows the image because I have `console.log(imgArray[curIndex])` and then it would show `after_response`.

<img width="444" alt="Screen Shot 2020-08-19 at 3 50 35 PM" src="https://user-images.githubusercontent.com/52719718/90707780-077c8800-e234-11ea-816c-f9937009bd33.png">

**Reply by vijaymarupudi** [reply]
I'm going to suggest a solution that does not use jsPsych's API, but instead uses Javascript's API.


```javascript
// As a state variable outside of the functions

let currentKeypressListener = null;

function slideShow() {
  // .. other logic

  if (currentKeypressListener) {
      document.removeEventListener('keypress', currentKeypressListener)
      currentKeypressListener = null
  }

  // after response replacement
  const listener = function(e) {
    if (trial.choices.includes(e.code)) {
      // ... keypress (after_response) logic
      document.removeEventListener('keypress', listener)
      currentKeypressListener = null
    }
  }
  currentKeypressListener = listener

  document.addEventListener('keypress', listener)
}
```

I think this should get rid of the problems. I think one of the issues was that you were adding a new listener for each loop, and they were stacking up. I'm not exactly sure why they were being called without a keypress, but I think the stacking makes it seem that many answers were acquired when a single keypress is being made.

**Reply by avashipman** [reply]
Thank you for your help!! It didn't record any of the responses I made but I might have written the code wrong. This is how I implemented the `after_response` variable and the Javascript API in the `slideShow` function:
  ```
 function slideShow() {
      console.log(imgArray[curIndex]);
      curIndex++;
      curIndex = curIndex % imgArray.length;
      document.getElementById("distractor-stimulus").classList.remove('responded')
      document.getElementById("distractor-stimulus").className = "fadeOut";
      setTimeout(function () {
        document.getElementById("distractor-stimulus").className = "";
        document.getElementById("distractor-stimulus").src = imgArray[curIndex];
      }, 800);

      //function to count the amount of time X appears
      function targetArray() {
        var x = imgArray[curIndex]
        if (x == "target.png") {
          targetData.push(x)
        }
      };
      targetArray();
      
      //keyboard listener
      if (currentKeypressListener) {
        document.removeEventListener('keypress', currentKeypressListener)
        currentKeypressListener = null
      }
      //after response replacement
      const listener = function(e) {
        if (trial.choices.includes(e.code)) {
          after_response
          document.removeEventListener('keypress', listener)
          currentKeypressListener = null
        }
      }
      currentKeypressListener = listener
      document.addEventListener('keypress', listener)
    };
    const intervalId = setInterval(slideShow, trial.response_next_image);
```

**Reply by vijaymarupudi** [reply]
Over here

```javascript
        after_response
        document.removeEventListener('keypress', listener)
```

You're not calling after_response, that line does nothing. I suggest changing after_response and adapting it to this new variant of the code. Note that you will not get an info object, unlike the getKeyboardResponse API, so you'll have to record the rt yourself. You can get the key code that was pressed using `e.code`. 

You can calculate rt yourself using [`performance.now()`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now). That's give you a timestamp in ms from when the page started loading. At the beginning of each slideShow() call, you'll want to set a variable, something like `var start = performance.now()`. And in the after_response section, you can calculate the rt by doing `var rt = performance.now() - start`.

**Reply by avashipman** [reply]
Ok, that is what I initially thought. I am not really concerned about the rt, but this is good information for the future! I tried just putting `distract_data.push(e.code)` in place of `after_response` to keep it simple, but this isn't working either. Also when I put `console.log(listener)` it does the same thing as the jsPsych API where it shows up every time even without a response.

**Reply by vijaymarupudi** [reply]
So one tangent I went one was figuring out that [`keyCode`] has been deprecated. They recommend to use `e.code` instead.

I went ahead and changed stuff until it seemed like it was working. `currentKeyboardListener` was not defined.

```diff
diff --git a/2_Stage copy.html b/2_Stage copy.html
index c0dd130..21bc063 100644
--- a/2_Stage copy.html	
+++ b/2_Stage copy.html	
@@ -9,6 +9,7 @@
   <script src="jspsych-6.1.0/plugins/jspsych-html-button-response.js"></script>
   <script src="jspsych-6.1.0/plugins/jspsych-image-keyboard-response.js"></script>
   <script src="jspsych-6.1.0/jspsych-survey-distractor-copy.js"></script>
+  <meta charset="utf-8">
   <link href="jspsych-6.1.0/css/jspsych.css" rel="stylesheet" type="text/css">
   <link href="jspsych-6.1.0/css/fadeOut.css" rel="stylesheet" type="text/css">
   <style>
@@ -88,7 +89,7 @@
             button_label: "SUBMIT",
             preamble: "Please write one sentence.",
             stimulus: jsPsych.randomization.sampleWithReplacement(images, 50, [1.5, 14, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 24]),
-            choices: [56],
+            choices: ['Digit8'],
             response_next_image: 1800,
             post_trial_gap: 500,
           },
@@ -128,7 +129,7 @@
             return [{ prompt: jsPsych.timelineVariable("prompt", true) }];
           },
           stimulus: jsPsych.timelineVariable("distractor"),
-          choices: [56],
+          choices: ['Digit8'],
           preamble: "Please write one sentence.",
           response_next_image: 1800,
           post_trial_gap: 500,
@@ -161,7 +162,7 @@
 
           var targetAmount = lastData.select('target').sum();
           var correctAmount = lastData.select('correctKeyPressed').values.toString();
-          var correct_amount = (correctAmount.match(/56/g) || []).length
+          var correct_amount = (correctAmount.match(/Digit8/g) || []).length
           console.log(correct_amount);
 
         return "<p> In this last section you typed "+wordCount+" words and detected "+correct_amount+" out of "+targetAmount+" X's.</p>" +
@@ -201,7 +202,7 @@
 
         var targetAmount = totalData.select('target').sum();
         var correctAmount = totalData.select('correctKeyPressed').values.toString();
-        var correct_amount = (correctAmount.match(/56/g) || []).length
+        var correct_amount = (correctAmount.match(/Digit8/g) || []).length
 
         var words = totalData.select('responses').values.toString();
         var wordsTotal = words.split(" ").length;
@@ -228,4 +229,4 @@
   </script>
 </body>
 
-</html>
\ No newline at end of file
+</html>
diff --git a/jspsych-6.1.0/jspsych-survey-distractor-copy.js b/jspsych-6.1.0/jspsych-survey-distractor-copy.js
index 7517c7d..b8c95ab 100644
--- a/jspsych-6.1.0/jspsych-survey-distractor-copy.js
+++ b/jspsych-6.1.0/jspsych-survey-distractor-copy.js
@@ -283,6 +283,8 @@ jsPsych.plugins["survey-distractor"] = (function () {
       }
     };
 
+    let currentKeypressListener = null;
+
     function slideShow() {
       console.log(imgArray[curIndex]);
       curIndex++;
@@ -311,6 +313,7 @@ jsPsych.plugins["survey-distractor"] = (function () {
       //after response replacement
       const listener = function(e) {
         if (trial.choices.includes(e.code)) {
+          console.log(e.code, trial.choices, distractor_data)
           distractor_data.push(e.code)
           document.removeEventListener('keypress', listener)
           currentKeypressListener = null
```

[`keyCode`]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

**Reply by avashipman** [reply]
This worked perfectly! Sorry for not catching that I should've changed the keyCode to e.code in the HTML. Thank you for all your help and patience!!

**Reply by vijaymarupudi** [reply]
Oh no, I learned about that today, I had no idea!

Glad to help :)
